### ISSUE TITLE
wrong code on valid code at O2+ (clang 14)

### ISSUE THREAD

**Issue by ecatmur** [llvm/llvm-project#56062]
```c++
int main() {
    struct D { struct P { int x, y; } a[2] = {}; } d;
    if (d.a[0].y != 0)
        throw &d.a[0].y;
}
```
```sh
$ clang++ --version
clang version 14.0.0
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/clang-14.0.0/bin
$ clang++ -O2 a.cpp && ./a.out
terminate called after throwing an instance of 'int*'
Aborted
```
clang 13 is not affected. According to godbolt it doesn't miscompile in clang trunk 15.0.0, but the bug is pretty fragile so I'm not sure that it's genuinely fixed. I'll try to bisect to find where it was introduced/removed.

**Comment by DimitryAndric** [comment]
At first sight, this looks to have been broken in 8cb9c7360944b1e4dd2e8125a48ede3628b459f1 (llvmorg-14-init-18064-g8cb9c7360944) "[LoopIdiom] Keep TBAA when creating memcpy/memmove", by @wsmoses, and subsequently fixed in e02f4976acbf086904bf3903348603cc0f04e6ad (llvmorg-15-init-6238-ge02f4976acbf) "[LoopIdiom] Merge TBAA of adjacent stores when creating memset", by @steplong & @efriedma-quic.

**Comment by DimitryAndric** [comment]
Also note that before 8cb9c7360944b1e4dd2e8125a48ede3628b459f1, the whole `throw` statement is optimized away:

```s
main:                                   # @main
        .cfi_startproc
# %bb.0:                                # %if.end
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        xorl    %eax, %eax
        popq    %rbp
        .cfi_def_cfa %rsp, 8
        retq
.Lfunc_end0:
        .size   main, .Lfunc_end0-main
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 14.0.0 (https://github.com/llvm/llvm-project.git d1daa54f36f0011eb621237e1f28ae7bc36f79fc)"
```

Afterwards, it actually throws:

```s
main:                                   # @main
        .cfi_startproc
# %bb.0:                                # %if.then
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        pushq   %rbx
        subq    $24, %rsp
        .cfi_offset %rbx, -24
        xorps   %xmm0, %xmm0
        movaps  %xmm0, -32(%rbp)
        leaq    -28(%rbp), %rbx
        movl    $8, %edi
        callq   __cxa_allocate_exception
        movq    %rbx, (%rax)
        movl    $_ZTIPi, %esi
        movq    %rax, %rdi
        xorl    %edx, %edx
        callq   __cxa_throw
.Lfunc_end0:
        .size   main, .Lfunc_end0-main
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 14.0.0 (https://github.com/llvm/llvm-project.git 8cb9c7360944b1e4dd2e8125a48ede3628b459f1)"
```

This continues until just before e02f4976acbf086904bf3903348603cc0f04e6ad, and from there, it optimizes the `throw` away again:

```s
main:                                   # @main
        .cfi_startproc
# %bb.0:                                # %if.then
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        pushq   %rbx
        subq    $24, %rsp
        .cfi_offset %rbx, -24
        xorps   %xmm0, %xmm0
        movaps  %xmm0, -32(%rbp)
        leaq    -28(%rbp), %rbx
        movl    $8, %edi
        callq   __cxa_allocate_exception
        movq    %rbx, (%rax)
        movl    $_ZTIPi, %esi
        movq    %rax, %rdi
        xorl    %edx, %edx
        callq   __cxa_throw
.Lfunc_end0:
        .size   main, .Lfunc_end0-main
        .cfi_endproc
                                        # -- End function
        .ident  "clang version 15.0.0 (https://github.com/llvm/llvm-project.git 907d3acefc3bdd6eb83f21589c6473ca7e88b3eb)"
```

It looks like it is able to prove the `if (d.a[0].y != 0)` condition can never be true, so it throws away the `throw`. :)

**Comment by efriedma-quic** [comment]
I think that fix was merged in 14.0.5.

**Comment by DimitryAndric** [comment]
> I think that fix was merged in 14.0.5.

Indeed, that was d350783a0520d09eb61e2eca1cd61c9cdac00908.

@ecatmur I think this can be considered solved, then? Please reopen if you still see similar issues.

**Comment by ecatmur** [comment]
Confirm that 14.0.5 works in the unreduced test case. Thanks!
